src/transformers/trainer_callback.py

```diff
@@ -475,3 +476,56 @@ def on_log(self, args, state, control, logs=None, **kwargs):
         _ = logs.pop("total_flos", None)
         if state.is_local_process_zero:
             print(logs)
+
+
+class EarlyStoppingCallback(TrainerCallback):
+    """
+        An early stopping callback
+
+        Args:
+           early_stopping_patience (:obj:`int`):
+                Use with :obj:`metric_for_best_model` to stop training when the specified metric worsens for
+                :obj:`early_stopping_patience` evaluation calls. early_stopping_threshold(:obj:`float`, `optional`):
+                Use with TrainingAruments :obj:`metric_for_best_model` and :obj:`early_stopping_patience` to denote how
+                much the specified metric must improve to satisfy early stopping conditions. `
+
+        Internal State: early_stopping_patience_counter (:obj:`int`): The number of times validation metrics failed to
+        improve.
+    """
+
+    def __init__(self, early_stopping_patience: int = 1, early_stopping_threshold: Optional[float] = 0.0):
+        self.training_bar = None
+        self.prediction_bar = None
+        self.early_stopping_patience = early_stopping_patience
+        self.early_stopping_threshold = early_stopping_threshold
+        self.early_stopping_patience_counter = 0
+
+    def check_metric_value(self, args, state, control, metric_value):
+        # best_metric is set by code for load_best_model
+        operator = np.greater if args.greater_is_better else np.less
+        if state.best_metric is None or (
+            operator(metric_value, state.best_metric)
+            and abs(metric_value - state.best_metric) > early_stopping_threshold
+        ):
+            self.early_stopping_patience_counter = 0
+        else:
+            self.early_stopping_patience_counter += 1
+
+    def on_evaluate(self, args, state, control, metrics, **kwargs):
+        # Ensure state.best_metric gets set in later steps
+        args.load_best_model_at_end = True
```

Comment:
This is the only drawback to arguments in the callback. Are we okay with this or is there a better way to make sure best_metric model arguments are enabled?

Reply:
The `TrainingArguments` should not be modified by a callback. I think we're only using `metric_for_best_model` here? maybe assert it's not None at `on_train_begin`?

Reply:
Fixed in https://github.com/huggingface/transformers/pull/8581/commits/7f43217e908576bad51b05fc02202246ca3dfd4b